libs/sdk-core/src/breez_services.rs

```diff
@@ -1305,57 +1302,71 @@ impl Receiver for PaymentReceiver {
+    /// could be invalid. Therefore, the [OpeningFeeParamsMenu] should always be initialized on-the-fly.
+    pub(crate) fn try_from(vals: Vec<OpeningFeeParams>) -> Result<Self> {
+        let temp = Self { vals };
+        temp.validate().map(|_| temp)
+    }
+
+    fn validate(&self) -> Result<()> {
+        match !self.vals.is_empty() {
+            true => {
+                // opening_fee_params_menu fees must be in ascending order
+                let is_ordered = self.vals.windows(2).all(|ofp| {
+                    ofp[0].min_msat < ofp[1].min_msat || ofp[0].proportional < ofp[1].proportional
+                });
+                // `valid_until` must not be a past datetime
+                let is_expired = self.vals.iter().all(|ofp| match ofp.valid_until_date() {
+                    Ok(valid_until) => Utc::now() > valid_until,
+                    Err(_) => {
+                        warn!("Failed to parse valid_until for OpeningFeeParams: {ofp:?}");
+                        false
+                    }
+                });
+                match is_ordered && !is_expired {
+                    true => Ok(()),
+                    false => Err(anyhow!("Failed to validate opening_fee_params_menu")),
+                }
+            }
+            false => Err(anyhow!(
+                "No channel opening params are available in lsp info"
+            )),
         }
     }
-}
 
-fn get_longest_valid_opening_fee_params(
-    opening_fee_params_menu: &Vec<OpeningFeeParams>,
-) -> Result<Option<OpeningFeeParams>> {
-    validate_opening_fee_params_menu(opening_fee_params_menu)?;
+    pub(crate) fn get_cheapest_opening_fee_params(&self) -> Option<OpeningFeeParams> {
+        self.vals.get(0).cloned()
+    }
 
-    // Find the fee params that are valid for at least 48h
-    let now = Utc::now();
-    let duration_48h = chrono::Duration::hours(48);
-    let mut valid_min_48h: Vec<OpeningFeeParams> = opening_fee_params_menu
-        .iter()
-        .filter(|f| f.valid_until - now > duration_48h)
-        .cloned()
-        .collect();
+    pub(crate) fn get_longest_valid_opening_fee_params(self) -> Result<Option<OpeningFeeParams>> {
+        // Find the fee params that are valid for at least 48h
+        let now = Utc::now();
+        let duration_48h = chrono::Duration::hours(48);
+        let mut valid_min_48h: Vec<OpeningFeeParams> = self
+            .vals
+            .into_iter()
+            .filter(|ofp| match ofp.valid_until_date() {
+                Ok(valid_until) => valid_until - now > duration_48h,
+                Err(_) => {
+                    warn!("Failed to parse valid_until for OpeningFeeParams: {ofp:?}");
+                    false
+                }
+            })
+            .collect();
 
-    // Of those, return the cheapest
-    valid_min_48h.sort_by_key(|f| f.min_msat);
-    Ok(valid_min_48h.first().cloned())
+        // Of those, return the cheapest
+        valid_min_48h.sort_by_key(|f| f.min_msat);
```

Comment:
I think this https://github.com/breez/breez-sdk/pull/282#discussion_r1244106104 was left out

> Note that these have to be sorted already according to the spec. So if the menu is valid, you shouldn't have to sort here again. Also, there could be multiple params with the same `min_msat`, but a different `proportional`.

The menu should already be sorted and lower min_msat doesn't necessarily mean cheaper fees as proportional fee may be much higher as explained.

Reply:
I addressed that and added a series of tests to cover all the sorting combos I could think of.

LMK if tests looks good, or if I missed anything.